sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -308,7 +399,10 @@ object CTESubstitution extends Rule[LogicalPlan] {
         p.copy(planBuilder = (nameParts, children) => {
           p.planBuilder.apply(nameParts, children) match {
             case u @ UnresolvedRelation(Seq(table), _, _) =>
-              resolveWithCTERelations(table, alwaysInline, cteRelations, u)
+              val resolved = resolveWithCTERelations(table, alwaysInline,
+                (recursiveCTERelation ++ cteRelations).toSeq, u)
+              recursionFound = resolved._2
```

Comment:
We have a problem here. This is a lambda function that is not executed immediately here. Seems we need to delay the recursive CTE resolution after all IDENTIFIER clauses are resolved, or we fail here first.

Reply:
Maybe we should revisit why we need to have `CTERelationDef#recursive`, or we update it later after seeing all its references.

Reply:
As we agreed, we won't support the case with identifier in recursive CTE. I added an exception here for that. The error is thrown in case recursiveCTERelation is not None. recursiveCTERelation set to be not None in resolveCTERelations in case allowRecursion is True (meaning we are under recursive CTE - `WITH RECURSIVE` specified). 

However, maybe it would be better and more clean to pass down the flag `allowRecursion` to `substituteCTE`. What do you think?